cpp/include/kvikio/file_handle.hpp

```diff
@@ -107,120 +126,121 @@ inline int open_fd(const std::string& file_path, const std::string& flags, mode_
+      _compat_mode{std::exchange(o._compat_mode, false)},
       _nbytes{std::exchange(o._nbytes, 0)},
       _handle{std::exchange(o._handle, CUfileHandle_t{})}
   {
   }
   FileHandle& operator=(FileHandle&& o) noexcept
   {
-    _fd     = std::exchange(o._fd, -1);
-    _own_fd = std::exchange(o._own_fd, false);
-    _closed = std::exchange(o._closed, true);
-    _nbytes = std::exchange(o._nbytes, 0);
-    _handle = std::exchange(o._handle, CUfileHandle_t{});
+    _fd_direct_on  = std::exchange(o._fd_direct_on, -1);
+    _fd_direct_off = std::exchange(o._fd_direct_off, -1);
+    _initialized   = std::exchange(o._initialized, false);
+    _compat_mode   = std::exchange(o._compat_mode, false);
+    _nbytes        = std::exchange(o._nbytes, 0);
+    _handle        = std::exchange(o._handle, CUfileHandle_t{});
     return *this;
   }
+  ~FileHandle() noexcept { close(); }
 
-  /**
-   * @brief FileHandle support move semantic but isn't copyable
-   */
-  ~FileHandle() noexcept
-  {
-    if (!_closed) { this->close(); }
-  }
-
-  [[nodiscard]] bool closed() const noexcept { return _closed; }
+  [[nodiscard]] bool closed() const noexcept { return !_initialized; }
 
   /**
-   * @brief Deregister the file and close the file if created with `steal_fd=true`
+   * @brief Deregister the file and close the two files
    */
   void close() noexcept
   {
-    _closed = true;
+    if (closed()) { return; }
 #ifdef KVIKIO_CUFILE_EXIST
-    if (!defaults::compat_mode()) { cuFileAPI::instance()->HandleDeregister(_handle); }
+    if (!_compat_mode) { cuFileAPI::instance()->HandleDeregister(_handle); }
 #endif
-    if (_own_fd) { ::close(_fd); }
+    ::close(_fd_direct_on);
+    ::close(_fd_direct_off);
+    _fd_direct_on  = -1;
+    _fd_direct_off = -1;
+    _initialized   = false;
   }
 
   /**
-   * @brief Get the file descripter of the open file
+   * @brief Get the file descripter of one of the open files
+   *
    * @return File descripter
    */
-  [[nodiscard]] int fd() const noexcept { return _fd; }
+  [[nodiscard]] int fd() const noexcept { return _fd_direct_off; }
 
   /**
-   * @brief Get the flags of the file descripter (see open(2))
+   * @brief Get the flags of one of the file descripters (see open(2))
+   *
    * @return File descripter
    */
-  [[nodiscard]] int fd_open_flags() const
-  {
-    /*NOLINTNEXTLINE(cppcoreguidelines-pro-type-vararg)*/
-    int ret = fcntl(_fd, F_GETFL);
-    if (ret == -1) {
-      throw std::system_error(errno, std::generic_category(), "Unable to retrieve open flags");
-    }
-    return ret;
-  }
+  [[nodiscard]] int fd_open_flags() const { return open_flags(_fd_direct_on); }
 
   /**
    * @brief Get the file size
    *
+   * The value are cached.
+   *
    * @return The number of bytes
    */
   [[nodiscard]] inline std::size_t nbytes() const
   {
-    if (_nbytes == 0 && _fd > 0) { _nbytes = get_file_size(_fd); }
+    if (_nbytes == 0) { _nbytes = get_file_size(_fd_direct_on); }
```

Comment:
Why is the file descriptor check dropped?

Reply:
Fixed